portfolio/src/main/webapp/script.js

```diff
@@ -129,24 +129,73 @@ function loadComments() {
   }
   const fetchString = `/data?maxcomments=${maxcomments}&metric=${sortMetric}&order=${sortOrder}`;
   fetch(fetchString).then(response => response.json()).then(comments => {
-    const commentList = document.getElementById("comments");
+    const commentList = document.getElementById("toplevelcomments");
     while (commentList.lastChild) {
       commentList.removeChild(commentList.lastChild);
     }
-    for (const comment of comments) {
-      commentList.appendChild(createListElement(comment));
+    const commentTree = locateChildren(comments);
+    let numDisplayed = 0;
+    for (commentId in commentTree) {
+      if (numDisplayed == maxcomments) {
+        break;
+      }
+      let comment = commentTree[commentId];
+      if (comment["parentId"] === 0) {
+        numDisplayed++;
+        commentList.appendChild(constructReplyTree(comment, commentTree, 40));
+      }
     }
+    commentList.style.marginLeft = "20px";
   });
 }
 
+function locateChildren(comments) {
+  let commentTree = {};
+  for (let i = 0; i < comments.length; i++) {
+    let id = comments[i]["id"];
+    commentTree[id] = comments[i];
+    commentTree[id]["children"] = [];
+    for (let j = 0; j < comments.length; j++) {
+      if (comments[j]["parentId"] === comments[i]["id"]) {
+        commentTree[id]["children"].push(comments[j]);
+      }
+    }
+  }
+  return commentTree;
+}
+
+function constructReplyTree(comment, commentTree, margin) {
+  let children = commentTree[comment["id"]]["children"];
+  if (children.length === 0) {
+    let thisReply = createListElement(comment);
+    return thisReply;
+  } else {
+    let thisReply = createListElement(comment);
+    let replyTree = document.createElement("ul");
+    replyTree.className = "comments";
+    const newMargin = margin + 20;
+    for (const child of children) {
+      const subTree = constructReplyTree(child, commentTree, newMargin);
+      replyTree.appendChild(subTree);
+    }
+    replyTree.style.marginLeft = `${margin}px`;
+    thisReply.appendChild(replyTree);
+    return thisReply;
+  }
+}
+
 // Creates a list element with the given comment text and metadata (name, timestamp etc.)
 function createListElement(comment) {
```

Comment:
[Optional] You might notice that it is tedious to do DOM manipulations in JS :) which is why most modern frontend frameworks try to help with this.

Nothing actionable here, but if you want you can play with some frontend frameworks (Angular, React, Vue, ...) to see how they make these code much easier to write and maintain.